### PR TITLE
ecs_service supports constraints and strategy update

### PR DESCRIPTION
**Pull request by zhoufeng1989**
##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

ecs_service

##### ADDITIONAL INFORMATION

related issue https://github.com/ansible-collections/community.aws/issues/1414

```paste below

```

### CODE REVIEW

**Comment by markuman** [comment]
@zhoufeng1989 looks basically good to me.  
Can you also append the `ecs_cluster` integration test please and verify that it works correctly?  
https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ecs_cluster/tasks/main.yml

**Comment by zhoufeng1989** [comment]
> @zhoufeng1989 looks basically good to me. Can you also append the `ecs_cluster` integration test please and verify that it works correctly? https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ecs_cluster/tasks/main.yml

Sure, will do it

**Comment by zhoufeng1989** [comment]
hi, @markuman , integration test cases are added, would you please have a look?  thanks~

**Comment by markuman** [comment]
@zhoufeng1989  can you apply those changes?

```diff
diff --git a/tests/integration/targets/ecs_cluster/tasks/main.yml b/tests/integration/targets/ecs_cluster/tasks/main.yml
index d105d087..7c81a346 100644
--- a/tests/integration/targets/ecs_cluster/tasks/main.yml
+++ b/tests/integration/targets/ecs_cluster/tasks/main.yml
@@ -108,19 +108,21 @@
         vpc_id: '{{ setup_vpc.vpc.id }}'
         rules: # allow all ssh traffic but nothing else
         - ports: 22
-          cidr: 0.0.0.0/0
+          cidr_ip: 0.0.0.0/0
       register: setup_sg
 
-    - name: find a suitable AMI
-      ec2_ami_info:
-        owner: amazon
-        filters:
-          description: "Amazon Linux AMI* ECS *"
-      register: ec2_ami_info
+    - set_fact:
+        # As a lookup plugin we don't have access to module_defaults
+        connection_args:
+          region: "{{ aws_region }}"
+          aws_access_key: "{{ aws_access_key }}"
+          aws_secret_key: "{{ aws_secret_key }}"
+          aws_security_token: "{{ security_token | default(omit) }}"
+      no_log: True
 
     - name: set image id fact
       set_fact:
-        ecs_image_id: "{{ (ec2_ami_info.images|last).image_id }}"
+        ecs_image_id: "{{ lookup('aws_ssm', '/aws/service/ecs/optimized-ami/amazon-linux-2/recommended/image_id', **connection_args) }}"
 
     - name: provision ec2 instance to create an image
       ec2_instance:
@@ -146,6 +148,10 @@
         modify_targets: no
         vpc_id: '{{ setup_vpc.vpc.id }}'
         target_type: instance
+        health_check_interval: 5
+        health_check_timeout: 2
+        healthy_threshold_count: 2
+        unhealthy_threshold_count: 2
       register: elb_target_group_instance
 
     - name: create second target group to use ip target_type
@@ -157,6 +163,10 @@
         modify_targets: no
         vpc_id: '{{ setup_vpc.vpc.id }}'
         target_type: ip
+        health_check_interval: 5
+        health_check_timeout: 2
+        healthy_threshold_count: 2
+        unhealthy_threshold_count: 2
       register: elb_target_group_ip
 
     - name: create load balancer
@@ -618,8 +628,8 @@
       assert:
        that:
           - ecs_service_creation_constraints.changed
-          - "{{ ecs_service_creation_constraints.service.placementConstraints | length }} == 1"
-          - "{{ ecs_service_creation_constraints.service.placementConstraints[0].type }} == distinctInstance"
+          - "ecs_service_creation_constraints.service.placementConstraints | length == 1"
+          - "ecs_service_creation_constraints.service.placementConstraints[0].type == 'distinctInstance'"
 
     - name: Update ecs service's placement constraints
       ecs_service:
@@ -642,9 +652,9 @@
       assert:
        that:
           - ecs_service_update_constraints.changed
-          - "{{ ecs_service_update_constraints.service.placementConstraints | length }} == 1"
-          - "{{ ecs_service_update_constraints.service.placementConstraints[0].type }} == memberOf"
-          - "{{ ecs_service_update_constraints.service.placementConstraints[0].expression }} == 'attribute:ecs.instance-type == t3.micro'"
+          - "ecs_service_update_constraints.service.placementConstraints | length == 1"
+          - "ecs_service_update_constraints.service.placementConstraints[0].type == 'memberOf'"
+          - "ecs_service_update_constraints.service.placementConstraints[0].expression == 'attribute:ecs.instance-type == t3.micro'"
 
     - name: Create ecs service with placement strategy
       ecs_service:
@@ -667,9 +677,9 @@
       assert:
        that:
           - ecs_service_creation_strategy.changed
-          - "{{ ecs_service_creation_strategy.service.placementplacementStrategy | length }} == 1"
-          - "{{ ecs_service_creation_strategy.service.placementConstraints[0].type }} == binpack"
-          - "{{ ecs_service_creation_strategy.service.placementConstraints[0].field }} == MEMORY"
+          - "ecs_service_creation_strategy.service.placementplacementStrategy | length == 1"
+          - "ecs_service_creation_strategy.service.placementConstraints[0].type == 'binpack'"
+          - "ecs_service_creation_strategy.service.placementConstraints[0].field == 'MEMORY'"
 
     - name: Update ecs service's placement strategy
       ecs_service:
@@ -692,9 +702,9 @@
       assert:
        that:
           - ecs_service_update_strategy.changed
-          - "{{ ecs_service_update_strategy.service.placementplacementStrategy | length }} == 1"
-          - "{{ ecs_service_update_strategy.service.placementConstraints[0].type }} == spread"
-          - "{{ ecs_service_update_strategy.service.placementConstraints[0].field }} == instanceId"
+          - "ecs_service_update_strategy.service.placementplacementStrategy | length == 1"
+          - "ecs_service_update_strategy.service.placementConstraints[0].type == 'spread'"
+          - "ecs_service_update_strategy.service.placementConstraints[0].field == 'instanceId'"
 
     # ============================================================
     # Begin tests for Fargate
```

File: changelogs/fragments/1601-ecs_service-support_constraints_and_strategy_update.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - ecs_service supports constraints and strategy update (https://github.com/ansible-collections/community.aws/pull/1601).
```
Comment (markuman) [inline changelogs/fragments/1601-ecs_service-support_constraints_and_strategy_update.yml:2]:
```suggestion
  - ecs_service  - respect ``placement_constraints`` for existing ecs services (https://github.com/ansible-collections/community.aws/pull/1601).
```

**Comment by zhoufeng1989** [comment]
hi @markuman, thanks for review, updated.

File: tests/integration/targets/ecs_cluster/tasks/main.yml [hunk 2]
```diff
@@ -598,6 +608,104 @@
...
+      register: ecs_service_update_strategy
+
+    - name: Assert ecs service strategy
+      assert:
+       that:
+          - ecs_service_update_strategy.changed
+          - "ecs_service_update_strategy.service.placementplacementStrategy | length == 1"
+          - "ecs_service_update_strategy.service.placementConstraints[0].type == spread"
+          - "ecs_service_update_strategy.service.placementConstraints[0].field == instanceId"
+
```
Comment (markuman) [inline tests/integration/targets/ecs_cluster/tasks/main.yml:708]:
```suggestion
    - name: Create ecs service with placement constraints
      ecs_service:
        name: "{{ ecs_service_name }}-constraint"
        cluster: "{{ ecs_cluster_name }}"
        load_balancers:
          - targetGroupArn: "{{ elb_target_group_instance.target_group_arn }}"
            containerName: "{{ ecs_task_name }}"
            containerPort: "{{ ecs_task_container_port }}"
        task_definition: "{{ ecs_task_name }}:{{ ecs_task_definition.taskdefinition.revision }}"
        scheduling_strategy: "REPLICA"
        placement_constraints:
          - type: distinctInstance
        desired_count: 1
        state: present
      register: ecs_service_creation_constraints

    - name: Assert ecs service constraint
      assert:
       that:
          - ecs_service_creation_constraints.changed
          - "ecs_service_creation_constraints.service.placementConstraints | length == 1"
          - "ecs_service_creation_constraints.service.placementConstraints[0].type == distinctInstance"

    - name: Update ecs service's placement constraints
      ecs_service:
        name: "{{ ecs_service_name }}-constraint"
        cluster: "{{ ecs_cluster_name }}"
        load_balancers:
          - targetGroupArn: "{{ elb_target_group_instance.target_group_arn }}"
            containerName: "{{ ecs_task_name }}"
            containerPort: "{{ ecs_task_container_port }}"
        task_definition: "{{ ecs_task_name }}:{{ ecs_task_definition.taskdefinition.revision }}"
        scheduling_strategy: "REPLICA"
        placement_constraints:
          - type: memberOf
            expression: 'attribute:ecs.instance-type == t3.micro'
        desired_count: 1
        state: present
      register: ecs_service_update_constraints

    - name: Assert ecs service constraint
      assert:
       that:
          - ecs_service_update_constraints.changed
          - "ecs_service_update_constraints.service.placementConstraints | length == 1"
          - "ecs_service_update_constraints.service.placementConstraints[0].type == memberOf"
          - "ecs_service_update_constraints.service.placementConstraints[0].expression == 'attribute:ecs.instance-type == t3.micro'"

    - name: Create ecs service with placement strategy
      ecs_service:
        name: "{{ ecs_service_name }}-strategy"
        cluster: "{{ ecs_cluster_name }}"
        load_balancers:
          - targetGroupArn: "{{ elb_target_group_instance.target_group_arn }}"
            containerName: "{{ ecs_task_name }}"
            containerPort: "{{ ecs_task_container_port }}"
        task_definition: "{{ ecs_task_name }}:{{ ecs_task_definition.taskdefinition.revision }}"
        scheduling_strategy: "REPLICA"
        placement_strategy:
          - type: binpack
            field: MEMORY
        desired_count: 1
        state: present
      register: ecs_service_creation_strategy

    - name: Assert ecs service strategy
      assert:
       that:
          - ecs_service_creation_strategy.changed
          - "ecs_service_creation_strategy.service.placementplacementStrategy | length == 1"
          - "ecs_service_creation_strategy.service.placementStrategy[0].type == binpack"
          - "ecs_service_creation_strategy.service.placementStrategy[0].field == MEMORY"

    - name: Update ecs service's placement strategy
      ecs_service:
        name: "{{ ecs_service_name }}-strategy"
        cluster: "{{ ecs_cluster_name }}"
        load_balancers:
          - targetGroupArn: "{{ elb_target_group_instance.target_group_arn }}"
            containerName: "{{ ecs_task_name }}"
            containerPort: "{{ ecs_task_container_port }}"
        task_definition: "{{ ecs_task_name }}:{{ ecs_task_definition.taskdefinition.revision }}"
        scheduling_strategy: "REPLICA"
        placement_strategy:
          - type: spread
            field: instanceId
        desired_count: 1
        state: present
      register: ecs_service_update_strategy

    - name: Assert ecs service strategy
      assert:
       that:
          - ecs_service_update_strategy.changed
          - "ecs_service_update_strategy.service.placementplacementStrategy | length == 1"
          - "ecs_service_update_strategy.service.placementStrategy[0].type == spread"
          - "ecs_service_update_strategy.service.placementStrategy[0].field == instanceId"

```

File: tests/integration/targets/ecs_cluster/tasks/main.yml [hunk 3]
```diff
@@ -598,6 +608,104 @@
...
+        state: present
+      register: ecs_service_update_strategy
+
+    - name: Assert ecs service strategy
+      assert:
+       that:
+          - ecs_service_update_strategy.changed
+          - "ecs_service_update_strategy.service.placementplacementStrategy | length == 1"
+          - "ecs_service_update_strategy.service.placementConstraints[0].type == spread"
+          - "ecs_service_update_strategy.service.placementConstraints[0].field == instanceId"
```
Comment (markuman) [inline tests/integration/targets/ecs_cluster/tasks/main.yml:707]:
```suggestion
        placement_strategy:
          - type: binpack
            field: MEMORY
        desired_count: 1
        state: present
      register: ecs_service_creation_strategy

    - name: Assert ecs service strategy
      assert:
       that:
          - ecs_service_creation_strategy.changed
          - "ecs_service_creation_strategy.service.placementplacementStrategy | length == 1"
          - "ecs_service_creation_strategy.service.placementStrategy[0].type == binpack"
          - "ecs_service_creation_strategy.service.placementStrategy[0].field == MEMORY"

    - name: Update ecs service's placement strategy
      ecs_service:
        name: "{{ ecs_service_name }}-strategy"
        cluster: "{{ ecs_cluster_name }}"
        load_balancers:
          - targetGroupArn: "{{ elb_target_group_instance.target_group_arn }}"
            containerName: "{{ ecs_task_name }}"
            containerPort: "{{ ecs_task_container_port }}"
        task_definition: "{{ ecs_task_name }}:{{ ecs_task_definition.taskdefinition.revision }}"
        scheduling_strategy: "REPLICA"
        placement_strategy:
          - type: spread
            field: instanceId
        desired_count: 1
        state: present
      register: ecs_service_update_strategy

    - name: Assert ecs service strategy
      assert:
       that:
          - ecs_service_update_strategy.changed
          - "ecs_service_update_strategy.service.placementplacementStrategy | length == 1"
          - "ecs_service_update_strategy.service.placementStrategy[0].type == spread"
          - "ecs_service_update_strategy.service.placementStrategy[0].field == instanceId"
```

this is a copy/paste mistake

**Comment by markuman** [comment]
Unfortunately CI is broken. But I've tested the new parameters with our grafana production environment.  
Imo, fixing the CI is out of scope from this PR. So let's ignore it from here.

Updating the service works fine for me.

But at the moment, the new parameter breaks backwards compatibility, because the default value for the placement strategy is 

```
TASK [debug] ********************************************************************************************************************************************************************
ok: [localhost] => {
    "ecs_service_creation_strategy.service.placementStrategy": [
        {
            "field": "attribute:ecs.availability-zone",
            "type": "spread"
        },
        {
            "field": "instanceId",
            "type": "spread"
        }
    ]
}
```

when nothing else is specified.

When you did this (_comment out the `placement_strategy` parameter and re-deploy your service_),

```diff
    - name: update or create ecs service for
      register: ecs_service_creation_strategy
      ecs_service:
        state: present
        region: "{{ region }}"
        name: "{{ SERVICE }}"
        cluster: "{{ CLUSTER }}"
        task_definition: "{{ SERVICE }}:{{ TASKREVISION }}"
        desired_count: "{{ DESIRED_COUNT }}"
        launch_type: EC2
        load_balancers: "{{ loadbalancers }}"
        deployment_configuration:
          minimum_healthy_percent: 50
+        # placement_strategy:
+        #   - type: binpack
+        #     field: MEMORY
```

I would expect, that the service is reverted to spread AZ strategy. But it stays in the binpack mode.  
So imo,. we must set defaults, no other placement strategy is given.  
Because when you initial deploy a service without given `placement_strategy`, it is deployed in AZ spread strategy.  
Can someone follow me? :)  
What do you think @alinabuzachis?

File: plugins/modules/ecs_service.py [hunk 4]
```diff
@@ -819,14 +819,22 @@ def create_service(self, service_name, cluster_name, task_definition, load_balan
...
             service=service_name,
             taskDefinition=task_definition,
             deploymentConfiguration=deployment_configuration)
+        # filter placement_constraint and left only those where value is not None
+        # use-case: `distinctInstance` type should never contain `expression`, but None will fail `str` type validation
+        if placement_constraints:
+            params['placementConstraints'] = [{key: value for key, value in constraint.items() if value is not None}
+                                              for constraint in placement_constraints]
+        if placement_strategy:
+            params['placementStrategy'] = placement_strategy
```
Comment (markuman) [inline plugins/modules/ecs_service.py:837]:
```suggestion
        # filter placement_constraint and left only those where value is not None
        # use-case: `distinctInstance` type should never contain `expression`, but None will fail `str` type validation
        if placement_constraints:
            params['placementConstraints'] = [{key: value for key, value in constraint.items() if value is not None}
                                              for constraint in placement_constraints]
        else:
            params['placementConstraints'] = []

        if placement_strategy:
            params['placementStrategy'] = placement_strategy
        else:
            params['placementStrategy'] = []
```

that solves the backwards compatiblity for me

File: tests/integration/targets/ecs_cluster/tasks/main.yml (hunk 3 above)
Comment (zhoufeng1989) [reply tests/integration/targets/ecs_cluster/tasks/main.yml]: Updated, sorry for slow response.

File: plugins/modules/ecs_service.py (hunk 4 above)
Comment (zhoufeng1989) [reply plugins/modules/ecs_service.py]: Updated, add test cases.

**Comment by markuman** [comment]
LGTM.
However, the integration test is broken. Fixing it is out of scope in thie PR.
Basically network is broken for instances that are deployed to the cluster.

File: changelogs/fragments/1601-ecs_service-support_constraints_and_strategy_update.yml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - ecs_service - respect ``placement_constraints`` for existing ecs services(https://github.com/ansible-collections/community.aws/pull/1601).
```
Comment (markuman) [inline changelogs/fragments/1601-ecs_service-support_constraints_and_strategy_update.yml:2]:
```suggestion
  - ecs_service - respect ``placement_constraints`` for existing ecs services (https://github.com/ansible-collections/community.aws/pull/1601).
```

File: plugins/modules/ecs_service.py [hunk 6]
```diff
@@ -819,14 +819,26 @@ def create_service(self, service_name, cluster_name, task_definition, load_balan
...
+        # use-case: `distinctInstance` type should never contain `expression`, but None will fail `str` type validation
+        if placement_constraints:
+            params['placementConstraints'] = [{key: value for key, value in constraint.items() if value is not None}
+                                              for constraint in placement_constraints]
+        else:
+            params['placementConstraints'] = []
+        if placement_strategy:
+            params['placementStrategy'] = placement_strategy
+        else:
+            params['placementStrategy'] = []
```
Comment (markuman) [inline plugins/modules/ecs_service.py:841]:
```suggestion
        if placement_constraints:
            params['placementConstraints'] = [{key: value for key, value in constraint.items() if value is not None}
                                              for constraint in placement_constraints]

        if placement_strategy:
            params['placementStrategy'] = placement_strategy

```

...ah that's my fault.

```
ansible-galaxy collection list|grep aws
amazon.aws                     5.1.0  
community.aws                  5.1.0  
```

I notice that the current module won't change the existing placements.

**Review by alinabuzachis — APPROVED** [review]

**Comment by markuman** [comment]
recheck

**Pull request merged**